File: legrid/priv/scripts/local_controller.py
#!/usr/bin/env python3
"""
Local LED Controller for Same-Machine Deployment

This script runs as a subprocess of the Elixir server and communicates
directly via stdin/stdout, eliminating WebSocket overhead.
"""

import sys
import time
import struct
import signal
import logging
import argparse
from typing import Dict, Tuple, List

# Add the py_controller directory to the path
sys.path.insert(0, "../../py_controller")

from hardware import create_hardware, MockHardware
from frame import FrameProcessor


class LocalLEDController:
    """Local LED controller that communicates via stdin/stdout"""

    def __init__(self, width: int, height: int, led_pin: int, led_count: int):
        self.width = width
        self.height = height
        self.led_pin = led_pin
        self.led_count = led_count

        # Set up logging
        logging.basicConfig(level=logging.INFO)
        self.logger = logging.getLogger("local-controller")

        # Initialize components
        self.running = True
        self.hardware = None
        self.frame_processor = None

        # Statistics
        self.frames_processed = 0
        self.last_frame_time = time.time()
        self.fps = 0.0

        # Set up signal handling
        signal.signal(signal.SIGINT, self._signal_handler)
        signal.signal(signal.SIGTERM, self._signal_handler)

        self.logger.info(
            f"Local LED Controller starting: {width}x{height}, {led_count} LEDs on pin {led_pin}"
        )

    def initialize(self):
        """Initialize hardware and frame processor"""
        try:
            # Create hardware interface
            config = {
                "width": self.width,
                "height": self.height,
                "led_count": self.led_count,
                "led_pin": self.led_pin,
                "brightness": 255,
                "layout": "serpentine",
                "flip_x": False,
                "flip_y": False,
                "transpose": False,
            }

            self.hardware = create_hardware(config)

            # Initialize frame processor
            self.frame_processor = FrameProcessor(
                width=self.width,
                height=self.height,
                layout="serpentine",
                flip_x=False,
                flip_y=False,
                transpose=False,
            )

            # Clear the display
            self.hardware.clear()

            self.logger.info("Local controller initialized successfully")
            return True

        except Exception as e:
            self.logger.error(f"Failed to initialize controller: {e}")
            return False

    def run(self):
        """Main run loop - read frames from stdin and display them"""
        if not self.initialize():
            return False

        self.logger.info("Local controller running. Reading frames from stdin...")

        try:
            while self.running:
                # Read frame data from stdin
                frame_data = self._read_frame()
                if frame_data is None:
                    continue

                # Process and display the frame
                self._process_frame(frame_data)

        except KeyboardInterrupt:
            self.logger.info("Keyboard interrupt received")
        except Exception as e:
            self.logger.error(f"Error in main loop: {e}")
        finally:
            self.cleanup()

        return True

    def _read_frame(self) -> bytes:
        """Read frame data from stdin"""
        try:
            # Read frame length (4 bytes, little-endian)
            length_bytes = sys.stdin.buffer.read(4)
            if not length_bytes or len(length_bytes) < 4:
                return None

            frame_length = struct.unpack("<I", length_bytes)[0]

            # Read frame data
            frame_data = sys.stdin.buffer.read(frame_length)
            if not frame_data or len(frame_data) < frame_length:
                return None

            return frame_data

        except Exception as e:
            self.logger.error(f"Error reading frame: {e}")
            return None

    def _process_frame(self, frame_data: bytes):
        """Process a frame and update the LEDs"""
        try:
            # Convert binary data to Frame object
            frame = self.frame_processor.process_binary_frame(frame_data)
            if not frame:
                return

            # Map logical frame to physical LED layout
            physical_pixels = self.frame_processor.map_led_layout(frame)

            # Update the hardware
            self._update_leds(physical_pixels)

            # Update statistics
            self._update_stats()

        except Exception as e:
            self.logger.error(f"Error processing frame: {e}")

    def _update_leds(self, pixels: List[Tuple[int, int, int]]):
        """Update the physical LEDs with pixel data"""
        if not pixels:
            return

        # Set each pixel
        for i, (r, g, b) in enumerate(pixels):
            if i < self.hardware.led_count:
                self.hardware.set_pixel(i, r, g, b)

        # Show the updated pixels
        self.hardware.show()

    def _update_stats(self):
        """Update frame statistics"""
        current_time = time.time()
        self.frames_processed += 1

        # Calculate FPS
        if self.last_frame_time > 0:
            delta_time = current_time - self.last_frame_time
            if delta_time > 0:
                instant_fps = 1.0 / delta_time
                self.fps = self.fps * 0.8 + instant_fps * 0.2

        self.last_frame_time = current_time

        # Send stats back to Elixir process periodically
        if self.frames_processed % 30 == 0:  # Every 30 frames
            self._send_stats()

    def _send_stats(self):
        """Send statistics back to the Elixir process"""
        stats = {
            "frames_processed": self.frames_processed,
            "fps": round(self.fps, 1),
            "hardware_type": "Mock"
            if isinstance(self.hardware, MockHardware)
            else "NeoPixel",
        }

        # Send stats as binary data
        stats_bytes = str(stats).encode("utf-8")
        length_bytes = struct.pack("<I", len(stats_bytes))

        try:
            sys.stdout.buffer.write(length_bytes)
            sys.stdout.buffer.write(stats_bytes)
            sys.stdout.buffer.flush()
        except Exception as e:
            self.logger.error(f"Error sending stats: {e}")

    def cleanup(self):
        """Clean up resources"""
        self.logger.info("Cleaning up...")

        if self.hardware:
            self.hardware.clear()
            self.hardware.cleanup()

        self.logger.info("Local controller shutdown complete")

    def _signal_handler(self, sig, frame):
        """Handle termination signals"""
        self.logger.info(f"Received signal {sig}")
        self.running = False


def main():
    """Main entry point"""
    parser = argparse.ArgumentParser(description="Local LED Controller")
    parser.add_argument("--width", type=int, default=25, help="Grid width")
    parser.add_argument("--height", type=int, default=24, help="Grid height")
    parser.add_argument("--led-pin", type=int, default=18, help="LED data pin")
    parser.add_argument(
        "--led-count", type=int, help="Number of LEDs (default: width*height)"
    )

    args = parser.parse_args()

    # Calculate LED count if not provided
    led_count = args.led_count or (args.width * args.height)

    # Create and run controller
    controller = LocalLEDController(
        width=args.width, height=args.height, led_pin=args.led_pin, led_count=led_count
    )

    success = controller.run()
    sys.exit(0 if success else 1)


if __name__ == "__main__":
    main()

File: legrid/priv/bin/local_controller.rs
use std::io::{self, Read, Write};
use std::time::{Duration, Instant};
use std::thread;

// LED control structures
#[derive(Debug, Clone, Copy)]
struct Pixel {
    r: u8,
    g: u8,
    b: u8,
}

struct LEDController {
    led_count: usize,
    pixels: Vec<Pixel>,
    frame_count: u64,
    last_frame_time: Option<Instant>,
    fps: f64,
}

impl LEDController {
    fn new(led_count: usize) -> Self {
        Self {
            led_count,
            pixels: vec![Pixel { r: 0, g: 0, b: 0 }; led_count],
            frame_count: 0,
            last_frame_time: None,
            fps: 0.0,
        }
    }

    fn process_frame(&mut self, frame_data: &[u8]) -> io::Result<()> {
        // Parse binary frame data
        if frame_data.len() < 10 {
            return Err(io::Error::new(io::ErrorKind::InvalidData, "Frame too short"));
        }

        // Parse header (version, type, frame_id, width, height)
        let width = u16::from_le_bytes([frame_data[6], frame_data[7]]);
        let height = u16::from_le_bytes([frame_data[8], frame_data[9]]);
        
        // Extract pixel data
        let pixel_data = &frame_data[10..];
        let expected_pixels = (width * height) as usize;
        
        if pixel_data.len() < expected_pixels * 3 {
            return Err(io::Error::new(io::ErrorKind::InvalidData, "Insufficient pixel data"));
        }

        // Convert to pixels
        for i in 0..expected_pixels.min(self.led_count) {
            let idx = i * 3;
            self.pixels[i] = Pixel {
                r: pixel_data[idx],
                g: pixel_data[idx + 1],
                b: pixel_data[idx + 2],
            };
        }

        // Update statistics
        self.frame_count += 1;
        let now = Instant::now();
        
        if let Some(last_time) = self.last_frame_time {
            let delta = now.duration_since(last_time).as_secs_f64();
            if delta > 0.0 {
                let instant_fps = 1.0 / delta;
                self.fps = self.fps * 0.8 + instant_fps * 0.2;
            }
        }
        
        self.last_frame_time = Some(now);

        // Send to hardware (mock implementation)
        self.send_to_hardware()?;

        Ok(())
    }

    fn send_to_hardware(&self) -> io::Result<()> {
        // Mock hardware implementation
        // In real implementation, this would control GPIO pins
        let lit_count = self.pixels.iter().filter(|p| p.r > 0 || p.g > 0 || p.b > 0).count();
        eprintln!("Frame {}: {}/{} pixels lit, FPS: {:.1}", 
                 self.frame_count, lit_count, self.led_count, self.fps);
        Ok(())
    }

    fn send_stats(&self) -> io::Result<()> {
        let stats = format!("{{\"frames_processed\":{},\"fps\":{:.1},\"hardware_type\":\"Rust\"}}", 
                           self.frame_count, self.fps);
        let stats_bytes = stats.as_bytes();
        let length = stats_bytes.len() as u32;
        
        // Send length (4 bytes, little-endian)
        io::stdout().write_all(&length.to_le_bytes())?;
        // Send stats
        io::stdout().write_all(stats_bytes)?;
        io::stdout().flush()?;
        
        Ok(())
    }
}

fn main() -> io::Result<()> {
    // Parse command line arguments
    let args: Vec<String> = std::env::args().collect();
    let mut width = 25;
    let mut height = 24;
    let mut led_pin = 18;
    let mut led_count = 600;

    for i in 1..args.len() {
        match args[i].as_str() {
            "--width" => {
                if i + 1 < args.len() {
                    width = args[i + 1].parse().unwrap_or(25);
                }
            }
            "--height" => {
                if i + 1 < args.len() {
                    height = args[i + 1].parse().unwrap_or(24);
                }
            }
            "--led-pin" => {
                if i + 1 < args.len() {
                    led_pin = args[i + 1].parse().unwrap_or(18);
                }
            }
            "--led-count" => {
                if i + 1 < args.len() {
                    led_count = args[i + 1].parse().unwrap_or(600);
                }
            }
            _ => {}
        }
    }

    eprintln!("Rust LED Controller starting: {}x{}, {} LEDs on pin {}", 
              width, height, led_count, led_pin);

    let mut controller = LEDController::new(led_count);
    let mut frame_count = 0;

    loop {
        // Read frame length (4 bytes, little-endian)
        let mut length_bytes = [0u8; 4];
        match io::stdin().read_exact(&mut length_bytes) {
            Ok(_) => {}
            Err(_) => break, // EOF or error
        }

        let frame_length = u32::from_le_bytes(length_bytes) as usize;
        
        // Read frame data
        let mut frame_data = vec![0u8; frame_length];
        match io::stdin().read_exact(&mut frame_data) {
            Ok(_) => {}
            Err(_) => break, // EOF or error
        }

        // Process frame
        if let Err(e) = controller.process_frame(&frame_data) {
            eprintln!("Error processing frame: {}", e);
            continue;
        }

        frame_count += 1;

        // Send stats periodically
        if frame_count % 30 == 0 {
            if let Err(e) = controller.send_stats() {
                eprintln!("Error sending stats: {}", e);
            }
        }
    }

    eprintln!("Rust LED Controller shutting down");
    Ok(())
} 